src/storage/src/source/reclock.rs

```diff
@@ -8,426 +8,327 @@
-            .boxed();
-
-        Ok(Self {
-            name,
-            read_progress,
-            write_upper,
-            persisted_timestamp_bindings,
-            read_cursors,
+        let mut operator = Self {
+            remap_trace: HashMap::new(),
+            since: as_of.clone(),
+            upper: Antichain::from_elem(Timestamp::minimum()),
+            source_upper: HashMap::new(),
+            listener: read_handle.listen(as_of.clone()).await.unwrap(),
             write_handle,
-            timestamp_bindings_listener,
+            read_handle,
             now,
-            update_interval,
-            last_timestamp: Timestamp::minimum(),
-        })
-    }
-
-    // TODO: require NowFn to be monotonic so we can simplify this function
-    fn get_time(&mut self) -> Option<Timestamp> {
-        let update_interval: u64 = self
-            .update_interval
-            .clone()
-            .as_millis()
-            .try_into()
-            .expect("interval millis don't fit into u64");
-        let new_ts = (self.now)();
-
-        // Mod and then round up
-        let remainder = new_ts % update_interval;
-        let new_ts_rounded = if remainder == 0 {
-            new_ts
-        } else {
-            new_ts - remainder + update_interval
+            update_interval_ms: u64::try_from(update_interval.as_millis()).unwrap(),
         };
 
-        if new_ts_rounded > self.last_timestamp {
-            self.last_timestamp = new_ts_rounded;
-            Some(new_ts_rounded)
-        } else {
-            None
-        }
+        operator.sync(&upper).await;
+
+        Ok(operator)
     }
 
-    // Ensure our persisted timestamp bindings go forward for a given partition.
-    fn validate_persisted_bindings(&self) -> bool {
-        for (partition, bindings) in self.persisted_timestamp_bindings.iter() {
-            let mut last_ts = Timestamp::minimum();
-            for (ts, _) in bindings.iter() {
-                assert!(
-                    *ts > last_ts,
-                    "TS GOES BACKWARDS: {:?} -> {:?} in PART {:?}; BINDINGS: {:?}",
-                    last_ts,
-                    ts,
-                    partition,
-                    self.persisted_timestamp_bindings
-                );
-                last_ts = *ts;
+    /// Reclocks the provided messages. The method returns an iterator that drains the hashmap
+    /// provided as it is being consumed.
+    pub async fn reclock<'a, M>(
+        &'a mut self,
+        messages: &'a mut HashMap<PartitionId, Vec<(M, MzOffset)>>,
+    ) -> ReclockIter<'a, M> {
+        // The caller provided *observed* messages so the upper is one past the highest offset
+        let mut source_upper = HashMap::new();
+        for (pid, part_messages) in messages.iter_mut() {
+            part_messages.sort_unstable_by(|a, b| a.1.cmp(&b.1));
+            if let Some((_, offset)) = part_messages.last() {
+                source_upper.insert(pid.clone(), *offset + 1);
```

Comment:
Can we add a comment somewhere saying that this function only works correctly so long as the min per-partition `MzOffset`s are non-decreasing?

Reply:
Hm, there are invariants that must be true and the code does try to maintain them (like not compacting too much) but preventing the min from decreasing is not something that the implementation guards against because there is no code path that descreases anyway (unlike compaction) . Did I completely miss what you're trying to say?